Add interoperation between `TranspileLayout` and `PropertySet`

jakelishman opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This adds a constructor and a "destructor" to `TranspileLayout` to cleanly centralise how the current ad-hoc `PropertySet` properties are combined into a `TranspileLayout`, and lets them be written back out again.  This lets passes in the middle of an execution pipeline construct a complete `TranspileLayout` structure, based on the current state.

This is a step along the road to centralising `TranspileLayout` into a core part of the IR; we can now add methods to it, write transpiler passes against `TranspileLayout`, and start transitioning the logic, without breaking backwards compatibility.  The idea is that passes will opt in to creating a `TranspileLayout`, mutate it, then write it back out again.  In the future, we can have the pass manager itself manage legacy passes by extracting the `TranspileLayout` into the `PropertySet` before executing a legacy pass.


### Details and comments

Based on #14825.

jakelishman committed:
Add interoperation between `TranspileLayout` and `PropertySet`

This adds a constructor and a "destructor" to `TranspileLayout` to
cleanly centralise how the current ad-hoc `PropertySet` properties are
combined into a `TranspileLayout`, and lets them be written back out
again.  This lets passes in the middle of an execution pipeline
construct a complete `TranspileLayout` structure, based on the current
state.

This is a step along the road to centralising `TranspileLayout` into a
core part of the IR; we can now add methods to it, write transpiler
passes against `TranspileLayout`, and start transitioning the logic,
without breaking backwards compatibility.  The idea is that passes will
opt in to creating a `TranspileLayout`, mutate it, then write it back
out again.  In the future, we can have the pass manager itself manage
legacy passes by extracting the `TranspileLayout` into the `PropertySet`
before executing a legacy pass.

jakelishman committed:
Remove unused import

jakelishman commented on qiskit/transpiler/layout.py:
```diff
@@ -738,3 +743,218 @@ def final_virtual_layout(self, filter_ancillas: bool = True) -> Layout:
+        # `qc.append(PermutationGate(permutation), qc.qubits)` to the output of the transpiler, were
+        # it to terminate at any given step, would be enough to precisely recreate the unitary of
+        # the circuit (with due hand-waviness around measurements/resets), up to the qubit
+        # relabelling of the initial layout.
```
This discussion probably actually wants to get moved into proper documentation in `qiskit.transpiler`, and the formalism completed.  Creating this documentation and the pedagogy of the rest of this function was actually one of the main points of the PR - it went hand-in-hand with creating the method itself, since I wanted to use both in the new Rust-y `ApplyLayout`, but I couldn't understand what everything meant.

alexanderivrii replied:
I find the above paragraph confusing. When you say "to the output of the transpiler", do you mean the circuit returned by the transpiler (without the output permutation, which is implicit), or the circuit with the output permutation (as above you argue that the transpiler really returns a triple that includes this output permutation as part of the implicit stuff)?

I presume you mean the circuit without the output permutation, and, under the assumption that the initial layout is trivial, appending the `PermutationGate(permutation)` gives us exactly the unitary operator of the original circuit (assuming it only consists of unitary operations). However I don't understand why this is called `"undoing"` rather than `"doing"`. The way I think about this say after the `ElidePermutationPass`, is that we have moved the explicit permutation gate at the end of the circuit to an implicit object, so to get the original operator we need to add it back (and not to undo it again).

I am fine keeping the word "undo", but could you please rephrase the paragraph to avoid any potential confusion?

jakelishman replied:
I did a bit more of a rewrite in 150d4a749c - hopefully that's a bit clearer.

> I presume you mean the circuit without the output permutation, and, under the assumption that the initial layout is trivial, appending the PermutationGate(permutation) gives us exactly the unitary operator of the original circuit.

Yes, but also in my world, you still do `qc.append(PermutationGate(permutation), qc.qubits)` first if you do have an initial layout, while the qubits are still the relabelled physical ones, _and then_ you undo the initial-layout relabelling to get back to the original unitary matrix.  The reason for it is that things that modify the final permutation shouldn't ever need to care what the original relabelling was (and they don't, in any current Qiskit pass).  In my 3-tuple form, they only deal with the `(_, explicit, implicit)` part, not the first `(relabelling, _, _)` bit, which means there's less to sync up.

> However I don't understand why this is called "undoing" rather than "doing". The way I think about this say after the ElidePermutationPass, is that we have moved the explicit permutation gate at the end of the circuit to an implicit object, so to get the original operator we need to add it back (and not to undo it again).

`ElidePermutations` is one of the few that reduces the number of swaps/permutations in the circuit explicitly at the moment.  Imo most things (like routing passes, which almost _always_ run) add swaps in that you'd need to "remove" to get back, which is why I used "undoing".  To me, you "undo" the effects of an elision by re-adding them whereas "doing" them again wouldn't imply an inverse.  But I tried to avoid using "undo" without greater context in the rewrite to help out.

alexanderivrii commented on qiskit/transpiler/layout.py:
```diff
@@ -738,3 +743,218 @@ def final_virtual_layout(self, filter_ancillas: bool = True) -> Layout:
+        # the way you think about permutations (it's not my preferred convention---Jake).
+        #
+        # Now, we'll step through the tranpsilation process.  At each point, we'll relate the
+        # objects we have back to a 3-tuple of abstract objects, which are applied in order:
```
Lol, even Jake occasionally misspells "transpile" 😄

jakelishman replied:
Man, I misspell `transpile` a good 50% of the time I type it...  Fixed in 150d4a749c5408b18c288057ddf826f66e035a3b.

alexanderivrii commented:
Thanks Jake for starting this. The following message is more of a high-level discussion/questions rather than a review.

First of all, I very-very much like the long-term vision described in this PR's summary. This is how I understand it (so please tell me if there are discrepancies between this and how you view it). Each of the transpiler passes takes a `DAGCircuit` as input, and produces a `DAGCircuit` as output. However, in the future, the `DAGCircuit` also explicitly contains a `TranspileLayout`. So we can think of this as a tuple `(explicit instructions in the DAG, TranspileLayout)`, and think of each pass as taking a tuple `(explicit instructions in the DAG, TranspileLayout)` as input, and producing a new tuple `(explicit instructions in the DAG, TranspileLayout)` as output. The important point being that such tuples alone are sufficient to fully describe the state of the current partially transpiled circuit, and in particular to construct a unitary operator in the case that of all the circuit instructions are unitary. For comparison, just looking at `explicit instructions in the DAG` alone does not suffice as this misses crucial information such as the mapping from the virtual qubits to the physical qubits and the final permutation of the output qubits (which are currently kept in the `PropertySet` separately from the `DAGCircuit`). In the code comments, this is actually described as a triple `(relabelling, explicit instructions, implicit instructions)`, where `explicit instructions` are the same as before while `relabeling` and `implicit instructions` are both part of the `TranspileLayout`. Jake, did I get this completely right, or your `implicit instructions` are more general than `TranspileLayout` (in particular, would the hypothetical "final clifford" be a part of `TranspileLayout` or of another object also sitting on the DAG)? The future `TranspileLayout` will be in rust and will be finalized by #14778, once we agree on all of the technical details.

One immediate question: this PR adds the interop between PropertySet and python-space TranspileLayout, while #14778 aims to define a Rust-space TranspileLayout. Which of these two PRs should be merged first? (Thinking through the interop here may have consequences as to how the rust-space TranspileLayout will eventually look like).

Currently, this PR replaces the `_finalize_layouts` method by 
```
layout := TranspileLayout.from_property_set(passmanager_ir, self.property_set)
# Write the canonicalized form back out. This is for backwards compatibility.
layout.write_into_property_set(self.property_set)
```
Currently this is done at the very end of the run and in particular canonicalizes away the horrible "virtual_permutation_layout" that we have introduced in the past, and from what I can judge this should work correctly. In addition, based on the release notes it seems that converting the property set into a transpile layout and back should also work after every pass in the transpilation pipeline, but I am confused as whether this already happens. I do want this to be the case after this PR is merged. There are two things that confuse me.
* Where in the transpiler pipeline, the associated (python-space) `TranspileLayout` is assumed/required to be `None`? For the rust-space, can we have `TranspileLayout` to be always non-None with `initial_layout` being an `Option` (I think I saw both your and Matthew comments to this extent)? Does going from property set to transpile layout and back (or from transpile layout to property set and back) work correctly when `TranspileLayout` is `None`?
* Most of the layout passes (e.g. `VF2Layout`) are analysis passes which compute how the mapping from virtual to physical qubits should look like, but do not yet apply this mapping. Should `TranspileLayout` know whether the layout has already been applied? (This would be easy with the Rust-space structure when it's an optional). Should `from_property_set` and `to_property_set` consider the cases "before the layout was applied" and "after the layout was applied" separately? Hmm, let's say we have just run the `VF2Analysis` pass and that we have a layout that we want to apply to the circuit. At this point the layout is not applied, so we do not want the `TranspileLayout` to change (which is consistent with `VF2Layout` being only an analysis pass). How should we keep the computed layout so that it does not get lost when converted from and to property set.

jakelishman commented:
> In the code comments, this is actually described as a triple (relabelling, explicit instructions, implicit instructions), where explicit instructions are the same as before while relabeling and implicit instructions are both part of the TranspileLayout. Jake, did I get this completely right, or your implicit instructions are more general than TranspileLayout (in particular, would the hypothetical "final clifford" be a part of TranspileLayout or of another object also sitting on the DAG)?

Yeah, this is correct with some subtlety: the `(relabelling, explicit, implicit)` 3-tuple is the abstract model of the transpilation state ("relabelling" might need some thought, but it's fine for now).  I wanted to define that outside a reference to any concrete objects.

Then we have to choose the concrete parts.  Obviously we have the DAG for "explicit".  We currently have a `Layout` object for "relabelling" (plus some extra metadata - bit order, etc), and a permutation for "implicit".  We then bundled all that up into a single object, `TranspileLayout`.

In my abstract model, we can choose to represent the 3 items however we like.  The actual choice of representation in any version of Qiskit is making a choice of the user API we offer, the backwards compatibility from the previous version, and the expressibility of the compiler.  We don't _have_ to keep `TranspileLayout`, but we probably want to do so and to evolve it in place to make backwards compatibility easier.

The "implicit" bit at the end is perhaps the most subtle.  We could store `implicit` as a separate DAG, and put whatever arbitrary operations we like on it (permutations, Cliffords, etc, etc).  The reason we don't want to do that is that it becomes near-impossible to derive anything useful from it after the fact, without running another compilation on it.  We deliberately choose to restrict what we track as "implicit" to make it easier to _use_ the output; it's always a trade-off between expression and ease-of-use.  I think there's a good argument to be made soon that a final Clifford, or the observable (as a Pauli measurement) of `EstimatorPub`, could and should also be considered in "implicit" in some way, and the abstract 3-tuple is supposed to be a way of formalising _why_ the transpiler is allowed to do it.  The mechanics of "how" are an API choice we then make.

> One immediate question: this PR adds the interop between PropertySet and python-space TranspileLayout, while https://github.com/Qiskit/qiskit/pull/14778 aims to define a Rust-space TranspileLayout. Which of these two PRs should be merged first? (Thinking through the interop here may have consequences as to how the rust-space TranspileLayout will eventually look like).

Matt and I (unfortunately) don't think we can merge the Python and Rust-space `TranspileLayout` objects together in a backwards-compatible way for the 2.x series, so they're going to co-exist.  After both merge, we'll have a PR that adds interop between the two classes as well, to enable Rust-space passes to start using Matt's class immediately too.


> Where in the transpiler pipeline, the associated (python-space) TranspileLayout is assumed/required to be None?

We can't really break backwards compatibility of this object, so for Qiskit 2.x, we're kind of stuck with the current state of affairs: if `virtual_permutation_layout` exists and something needs to examine the "layout", it'll see a dummy (and incorrect) `initial_layout` too.  We'll fix that in Qiskit 3.0.  If neither `virtual_permutation_layout` and `initial_layout` are set, this object is `None`.

> Most of the layout passes (e.g. VF2Layout) are analysis passes which compute how the mapping from virtual to physical qubits should look like, but do not yet apply this mapping. Should TranspileLayout know whether the layout has already been applied?

Once we transition to `TranspileLayout` being on `DAGCircuit` directly, there'll _still_ be a `layout` key in the `PropertySet`, so `VF2Layout` can still be an analysis pass.  `ApplyLayout` will be able to read in the property-set keys to know what layout it needs to apply, and will be able to examine the `DAGCircuit` (via its `TranspileLayout` attribute) to know how to apply the futher remapping.

jakelishman commented:
The short version of it:

This PR isn't intended to change any behaviour in Qiskit 2.x.  It's just laying the groundwork of the conversion functions that will be necessary in the upgrade path to Qiskit 3.x.

mtreinish commented on qiskit/transpiler/layout.py:
```diff
@@ -738,3 +743,218 @@ def final_virtual_layout(self, filter_ancillas: bool = True) -> Layout:
+                In all standard use, this is a required field.  However, if
+                ``virtual_permutation_layout`` is set, then a "trivial" layout will be inferred,
+                even if the circuit is not actually laid out to hardware.  This is an unfortunate
+                limitation of the circuit data model, where it is not possible to specify a final
```
Is it a limit of the circuit data model or the `TranspileLayout` data model?

jakelishman replied:
`TranspileLayout` data model more, though at the time I wrote the comment, I wasn't yet making a proper distinction, and there's still complications that we assume elsewhere that `QuantumCircuit.layout is not None` implies "laid out to hardware" because of the `TranspileLayout` data-model deficiency.

jakelishman replied:
Changed in 3dbf5f182a.

mtreinish commented on qiskit/transpiler/layout.py:
```diff
@@ -738,3 +743,218 @@ def final_virtual_layout(self, filter_ancillas: bool = True) -> Layout:
+                permutation without also having an initial layout.
+
+        ``original_qubit_indices``
+            **Required** (but automatically set by the :class:`.PassManager`).  The mapping
```
Just checking this resolves to `qiskit.transpiler.PassManager` and not `qiskit.passmanager.PassManager`. I assume it does because of the reverse resolution order. But figured I'd ask because it's the kind of thing that's easy to slip.

jakelishman replied:
The one in `qiskit.passmanager` is called `BasePassManager`.

jakelishman replied:
Sphinx doesn't let you use the prefix-`.` lookup if there's more than one possible match - it's a warning/error (either way, it fails our build).

mtreinish commented on qiskit/transpiler/layout.py:
```diff
@@ -738,3 +743,218 @@ def final_virtual_layout(self, filter_ancillas: bool = True) -> Layout:
+            virtual_permutation_layout[virtual_bit]
+            for virtual_bit in input_qubits[:num_input_qubits]
+        ]
+        # `virtual_permutation_layout` is defined without ancillas.  If they got added later, extend
+        # the virtual permutation with the implicit identity on the other components.
+        if num_qubits > len(undo_elided_on_virtuals):
+            undo_elided_on_virtuals.extend(range(len(undo_elided_on_virtuals), num_qubits))
```
Did this happen in your testing? I'd be surprised if so, because we should expand with ancillas after anything that introduces a virtual permutation and I would expect the ancilla expansion or layout application to transform the property set.

jakelishman replied:
> I would expect the ancilla expansion or layout application to transform the property set.

That doesn't happen (and is one of the reasons I _really_ want this to move to the DAG and to document that passes _must_ update it when mutating the DAG).  Amazingly, I think in the parent of this PR, it's still `ApplyLayout` that "expands" `original_qubit_indices` instead of, you know, the pass that did the expansion.

mtreinish replied:
Ah sorry it was a typo, I meant to say I **wouldn't** expect. That matches my expectation and memory of how those passes work.

jakelishman replied:
Oh right I see - in that case:

> we should expand with ancillas after anything that introduces a virtual permutation

yeah, this is why we need the logic.  The same stuff exists in the version of `PassManager._finalize_layouts` in Qiskit 2.1.  We want to hold a permutation that's the same width as the circuit, but once ancilla expansion has happened, `virtual_permutation_layout` is still only the length it was before expansion, so it's too short.

mtreinish commented on releasenotes/notes/transpilelayout-from_property_set-97eb505156c13230.yaml:
```diff
@@ -0,0 +1,14 @@
+    :class:`.TranspileLayout` will be, modify it in the fully structured form, and then write it
+    back out in canonical form.
+
+    It is expected that in a future major version of Qiskit, the :class:`.TranspileLayout`
```
You didn't want to say 3.0 here?

jakelishman replied:
Done in 3dbf5f182a.

mtreinish reviewed (commented):
Overall this LGTM, it's really nice to have a clear description of all the moving pieces here and how it all fits together. I agree with your inline comment we should bubble this up to the transpile module docs to make it clear for people (especially as we now have docs for pass authors) I just had a few small questions inline but otherwise I think I"m happy to merge this.

alexanderivrii reviewed (commented):
This looks good to me as well. Thanks for adding the detailed description of what each field in the property set means, our assumptions about them, and how they interact which each other. I have asked to rephrase one of the paragraphs, since I believe there might be still room for confusion left.

One other question. This PR allows to "canonicalize" the property set by calling `from_property_set` followed by `write_into_property_set`. In practice, this PR calls these once at the very end of transpilation instead of `_finalize_layouts`, where I believe everything works perfectly. Hence I am happy to merge this as is. However, is it true that we can currently "canonicalize" the property set after every transpiler pass and we would still get correct results? 

I am afraid this might not be true (for instance, would this work right after `ElidePermutations` or after one of the analysis passes the compute but do not apply the layout?), I would be happy to be wrong, but if not - can you please add an explicit warning to this extent?

alexanderivrii commented:
After some experimenting, I can indeed see that "canonicalizing" the transpile layout right after the elide permutations pass leads to incorrect results. This is expected (due to the deficiency of our python-based transpile layout class) and the reason behind introducing the (arguably very ugly) `virtual_permutation_layout` in the first place. This should not affect #14904 because when `ApplyLayout` finally runs we have full information (including the mapping from virtual to physical qubits) to be able to transfer the permutation from virtual to physical qubits. I don't know if we can do anything with this, as the Python-based TranspileLayout class does not have `virtual_permutation_layout`, so we should probably just document this clearly in the code that canonicalizations works after `ApplyLayout` or `SabreLayout` (when that explicitly calls `ApplyLayout`). What do you think?

jakelishman commented:
Sasha: can you give a code example showing that this is wrong after `ElidePermutations`?

alexanderivrii commented:
Jake:

The following code (with your PR applied)
```
qc = QuantumCircuit(3)
qc.h(0)
qc.swap(0, 2)
qc.cx(0, 2)
qc.swap(1, 0)
qc.h(1)
print(qc.draw(idle_wires=True))

coupling_map = CouplingMap([(0, 1), (1, 0), (1, 2), (2, 1)])
target = Target.from_configuration(basis_gates=["u", "cx"], num_qubits=coupling_map.size(), coupling_map=coupling_map)

# the coupling map + the initial layout ensure that we will have to deal both with the permutation of virtual qubits
# (virtual permutation layout) and with the mapping from virtual to physical qubits (initial_layout)
pm = generate_preset_pass_manager(optimization_level=3, seed_transpiler=42, target = target, initial_layout=[1, 2, 0])
qct = pm.run(qc)
print(qct.draw(idle_wires=True))

eq = (Operator.from_circuit(qct).equiv(Operator(qc)))
print(f"{eq =  }")
```
prints `True` (as it should).

Adding
```
from qiskit.transpiler.layout import TranspileLayout 
if (layout := TranspileLayout.from_property_set(new_dag, self.property_set)) is not None:
    layout.write_into_property_set(self.property_set)
```
to the `ElidePermutations` pass just before returning `new_dag` outputs `False`.

jakelishman commented:
Have you got #14919 included in your branch when you try that?

alexanderivrii commented:
Not explicitly (let me try).

Update: yeah unfortunately the problem persists even with #14919 applied (I was pretty sure this would happen, but wanted to double-check).

jakelishman commented:
Thanks, I'll look into it - it might be because of the effective double-set of the initial layout (due to how we have to handle it with `ElidePermutations` right now).

jakelishman committed:
Be more commital about 3.0

jakelishman committed:
Make comment more explicit about state

jakelishman commented:
As to Sasha's comment above about the potential bug (https://github.com/Qiskit/qiskit/pull/14826#issuecomment-3209777695): actually this is because our logic in the default layout plugin construction is pretty incompatible with `layout` or `final_layout` being set before the layout plugin runs.  We can update that in a follow-up, when we want to start using this method in earlier passes, I think.  (*edit*: actually there _is_ a bug in this PR currently in that case too, but I'll fix it tomorrow now.)

There is a different problem with using the `TranspileLayout.write_into_property_set` normalisation indiscriminately, though, which is that if a pass that sets `virtual_permutation_layout` does it, and then _another_ pass that also updates `virtual_permutation_layout` runs, the first pass's permutation will have been normalised into `final_layout`, which the normalisation assumes got filled in _later_ than `virtual_permutation_layout`, which will result in incorrect results.  I don't see a neat way around this (other than adding extra temporary tracking crap to the property set), and since (I'm pretty sure) all passes that set `virtual_permutation_layout` already fail to update `virtual_permutation_layout` in a consistent way, I think we can get away with it.  We just need to be careful in the upgrade path, but we were always going to need to take care with the ordering of updates.

jakelishman committed:
Handle case of `final_layout` without `layout`

If a virtual-permutation setting pass uses
`TranspileLayout.write_into_property_set` to normalise its virtual
permutation into a final layout, we would have subsequently failed to
detect the permutation if the initial layout was (validly) reset to
`None`.  This corrects the logic.

jakelishman commented:
Ok, so to speak to Sasha's bug, there's about 3 things going on:
- the most immediate problem is that the preset-passmanager logic uses `final_layout` as a proxy for "did we reach SabreLayout+Routing and the routing bit ran?". If so, it doesn't run `ApplyLayout`.  Normalising into `final_layout` tricks that conditional.  We can tighten up that check later, I think.
- this PR alone doesn't make `ApplyLayout` use this code yet (you need #14904 for that), and that makes `ApplyLayout` properly aware of how to normalise things.
- if you deal with both of those things, there was a bug in this logic where it wouldn't handle the case of `(initial_layout, virtual_permutation_layout, final_layout)` matching `(None, None, not-None)`.  That's fixed in 30f40de012.

With all three things fixed, Sasha's example with the modified `ElidePermutations` works correctly for me.

I think there'll be a tail of issues to work through as we transition the transpiler pipeline over to using `TranspileLayout` on `DAGCircuit` itself - I think I've done everything we can do in this part of the issue now, though.

mtreinish reviewed (approved):
This LGTM now, thanks for the updates.
